java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+        String prefix = "channel.";
+        String pattern = prefix + "*";
+        int numChannels = 256;
+        int messagesPerChannel = 256;
+        ChannelMode mode = standalone ? PubSubChannelMode.PATTERN : PubSubClusterChannelMode.PATTERN;
+        var messages = new ArrayList<Message>(numChannels * messagesPerChannel);
+        var subscriptions = Map.of(mode, Set.of(pattern));
+
+        for (var i = 0; i < numChannels; i++) {
+            var channel = prefix + UUID.randomUUID();
+            for (var j = 0; j < messagesPerChannel; j++) {
+                var message = UUID.randomUUID().toString();
+                messages.add(new Message(message, channel, pattern));
+            }
+        }
+
+        MessageCallback callback =
+                (msg, ctx) -> ((ConcurrentLinkedDeque<Pair<Integer, Message>>) ctx).push(Pair.of(1, msg));
+
+        var listener =
+                useCallback
+                        ? createClientWithSubscriptions(
+                                standalone, subscriptions, Optional.of(callback), Optional.of(mq))
+                        : createClientWithSubscriptions(standalone, subscriptions);
+        var sender = createClient(standalone);
+        clients.addAll(List.of(listener, sender));
+
+        Thread.sleep(404); // need some time to propagate subscriptions - why?
+
+        for (var message : messages) {
+            // TODO why publish returns 0 on cluster? meanwhile messages are delivered
+            var res = sender.publish(message.getChannel(), message.getMessage()).get();
+            if (standalone) {
+                assertEquals(1L, res);
+            }
+        }
+        assertEquals(0L, sender.publish("channel", UUID.randomUUID().toString()).get());
+        Thread.sleep(404); // deliver the messages
+
+        verifyReceivedMessages(
+                messages.stream().map(m -> Pair.of(1, m)).collect(Collectors.toSet()),
+                listener,
+                useCallback);
+    }
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = "standalone = {0}, use callback = {1}")
+    @MethodSource("getTwoBoolPermutations")
+    public void combined_exact_and_pattern_one_client(boolean standalone, boolean useCallback) {
+        String prefix = "channel.";
+        String pattern = prefix + "*";
+        int numChannels = 256;
+        int messagesPerChannel = 256;
+        var messages = new ArrayList<Message>(numChannels * messagesPerChannel);
+        ChannelMode mode = standalone ? PubSubChannelMode.EXACT : PubSubClusterChannelMode.EXACT;
+        Map<? extends ChannelMode, Set<String>> subscriptions =
+                Map.of(
+                        mode,
+                        new HashSet<>(),
+                        standalone ? PubSubChannelMode.PATTERN : PubSubClusterChannelMode.PATTERN,
+                        Set.of(pattern));
+
+        for (var i = 0; i < numChannels; i++) {
```

Comment:
This nested loop also shows up in alot of tests. Refactor to GenerateChannelsAndMessages() that takes in output channel and message lists.